Add debug widget into 3D view

Withalion opened the pull request:
## Description

The new debug side panel moves all relevant debug settings into one place, from multiple into one easily accessible side 
panel. Furthermore the change implemented in #58947 was missing any option to toggle this on/off. 

Features in this PR:
- New checkbox for stopping scene updates
- Options moved from advanced sub-category
- Camera info moved from navigation side panel and changed the way it's shown (changing values through the new spin boxes in next PR)
- Debug depth map view default position moved from top right corner into bottom left, so it won't collide with the axis view
- The size of toolbar icons is responsive to the docked state of window  

Standalone window:
![Screenshot_20241118_224841](https://github.com/user-attachments/assets/5b4f8db9-1c16-4aa9-a5a2-cc730511b67c)

Docked window:
![Screenshot_20241118_224929](https://github.com/user-attachments/assets/475b515f-5b61-4b29-8669-11a66df75336)


<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

Withalion committed:
Add stop updates checkbox in 3D View options

Withalion committed:
Connect stop updates setting to the scene

Currently this bricks the screen render

Withalion committed:
[feature] Create new debug widget for 3D map View

This commit adds new debug side panel (similar to navigation widget), which moves several existing options from settings/navigation panel here. Camera Info has been moved from navigation panel and from table changed into input elements (preparation for another feature). Show map tile info, Show bounding boxes, Show camera's view/rotation center, Show light sources, Show debug overlay, Debug shadow map, Debug Depth map options have been moved from advanced sub-category of options. New Stop scene updates option has been created.

Withalion committed:
Connect debug widget to Qgs3dMapSettings

Withalion committed:
Add minor UI improvements

Debug camera info is always shown
Debug depth map widget's default position is moved to bottom left
corner, so it won't collide with axis widget
Docking icon is moved to the last position
Toolbar icon size is finally correct, changes with the window docked state as
it should

nyalldawson commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -305,6 +308,19 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
 
   onTotalPendingJobsCountChanged();
 
+  QAction *toggleDebugPanel = toolBar->addAction(
```
I don't think we should make this visible by default. It's not something the average user ever needs, and adding it to the toolbar just contributes to UI clutter. Instead of a toolbar action, how about some hidden key shortcut to toggle the pane?

uclaros replied:
I agree, but I'd rather have a `Show debug panel` checkbox in advanced settings page instead of completely hiding it!

Withalion replied:
I like the @uclaros approach of having a checkbox + I can add a keyboard shortcut

Withalion replied:
I moved the setting into advanced sub-category of options and added a Ctrl+Shift+d shortcut for the panel

nyalldawson replied:
@Withalion 
> Ctrl+Shift+d 

That's already used for opening Data Source Manager. How about F12? That's used for the dev tools panel in the main window, so I think it'd be nice to use that here too. (I realise that if the 3d view is docked then F12 would only work if the 3d view is focused, but that same behavior would also be present with Ctrl+Shift+D due to the conflict with the Data Source Manager shortcut)

Withalion replied:
Okay that makes sense. I guess having an overlap in shortcuts is unavoidable, but it's definitely better UX on F12.

Withalion replied:
@nyalldawson after testing the change, F12 on docked windows doesn't work at all. For the previous combination I at least got a window about shortcut ambiguity, but on F12 I don't get anything. It just doesn't work. In stand alone window it works.

Withalion committed:
Move debug panel icon from toolbar

Debug panel in 3d map view can be set in advanced settings or by
keyboard shortcut CTRL + SHIFT + d

Withalion committed:
Fix code to best practices

Withalion commented:
Regarding the checkboxes, yes totally agree however that's the theme issue. I'm using the default theme in dark mode, other themes don't have this issue. Of course the issue is in the whole app not just in the debug panel

wonder-sk commented on src/3d/qgs3dmapsettings.cpp:
```diff
@@ -70,6 +70,7 @@ Qgs3DMapSettings::Qgs3DMapSettings( const Qgs3DMapSettings &other )
   , mShowCameraRotationCenter( other.mShowCameraRotationCenter )
   , mShowLightSources( other.mShowLightSources )
   , mShowLabels( other.mShowLabels )
+  , mStopUpdates( other.mStopUpdates )
```
probably also `mShowDebugPanel` should be copied

wonder-sk commented on src/3d/qgs3dmapsettings.cpp:
```diff
@@ -313,6 +314,7 @@ void Qgs3DMapSettings::readXml( const QDomElement &elem, const QgsReadWriteConte
   mShowCameraRotationCenter = elemDebug.attribute( QStringLiteral( "camera-rotation-center" ), QStringLiteral( "0" ) ).toInt();
   mShowLightSources = elemDebug.attribute( QStringLiteral( "show-light-sources" ), QStringLiteral( "0" ) ).toInt();
   mIsFpsCounterEnabled = elemDebug.attribute( QStringLiteral( "show-fps-counter" ), QStringLiteral( "0" ) ).toInt();
+  mStopUpdates = elemDebug.attribute( QStringLiteral( "stop-updates" ), QStringLiteral( "0" ) ).toInt();
```
probably also `mShowDebugPanel` should be read from / written to xml...

wonder-sk commented on src/3d/qgs3dmapsettings.h:
```diff
@@ -431,6 +431,14 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
+    */
+    void setStopUpdates( bool enabled );
+    //! Returns whether to keep updating the scene on zoom
+    bool stopUpdates() const;
```
ideally this method should be also marked with `\since QGIS 3.42`

wonder-sk commented on src/3d/qgs3dmapsettings.h:
```diff
@@ -431,6 +431,14 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
     int eyeDomeLightingDistance() const;
 
+    /**
+    * Sets whether to keep updating scene on zooming or keep the same amount of points visible
```
this "stop updates" does not apply only to points, but pretty much all 3D entities - terrain, vector layers, tiled scene layers etc. So best to rephrase this to something like "Sets whether scene updates upon camera movement should be enabled", maybe even add a note that by default scene always gets updated and this option is good for debugging purposes.

wonder-sk commented on src/app/3d/qgs3ddebugwidget.cpp:
```diff
@@ -0,0 +1,130 @@
+#include "qgs3dmapcanvas.h"
+#include "qgscameracontroller.h"
+
+Qgs3DDebugWidget::Qgs3DDebugWidget( Qgs3DMapCanvas *canvas, QWidget *parent ) : QWidget( parent ), m3DMapCanvas( canvas )
```
it is a habit in qgis code to put the initializer list on the next line, and have each member initialized on a separate line for better readability

wonder-sk commented on src/app/3d/qgs3ddebugwidget.cpp:
```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Sets up the interactive elements with values from Qgs3DMapSettings
+ */
```
Common practice in QGIS code is to have function docstrings in the header... (also for updateFromCamera())

wonder-sk commented on src/app/3d/qgs3ddebugwidget.h:
```diff
@@ -0,0 +1,41 @@
+class Qgs3DMapCanvas;
+
+#include <ui_3ddebugwidget.h>
+#include "qgs3dmapsettings.h"
```
we don't need to include qgs3dmapsettings.h here, only in the cpp - here we can have just forward declaration

Withalion replied:
compiler doesn't agree on this :( so it will have to stay

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -288,12 +289,18 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
   hLayout->addWidget( mNavigationWidget );
+  hLayout->addWidget( mDebugWidget );
 
+  auto *debugPanelShortCut = new QShortcut( QKeySequence( tr( "Ctrl+Shift+d" ) ), this );
```
It is a habit in QGIS code to only use `auto` for complex types (e.g. iterators) - for "simple" types it is preferred to write the type for code clarity (arguably here's it is clear, but in other places it may not be as clear, e.g. when assigning returned value from a function)

wonder-sk commented on src/app/3d/qgs3ddebugwidget.cpp:
```diff
@@ -0,0 +1,130 @@
+  scrollAreaWidgetContents->adjustSize();
+  scrollArea->setMinimumWidth( scrollArea->sizeHint().width() );
+  scrollArea->adjustSize();
+  this->adjustSize();
```
```suggestion
  adjustSize();
```
(best to avoid `this->` unless it is really needed)

wonder-sk reviewed (commented):
Looking good - finally we're on the way to proper developer tools for 3D scenes :slightly_smiling_face:

Withalion committed:
Fix issues found in review

Withalion committed:
Move documentation into header file
